sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -36,75 +38,345 @@ class GpuGenerateExecSparkPlanMeta(
+    // how may splits will we need to keep exploding working safely
+    val numSplits = numSplitsForTargetSize max numSplitsForTargetRow
+
+    if (numSplits == 0) Array()
+    else GpuBatchUtils.generateSplitIndices(inputRows, numSplits)
+  }
+
+  // Infer result schema of GenerateExec from input schema
+  protected def resultSchema(inputSchema: Array[DataType],
+    genOffset: Int,
+    includePos: Boolean = false): Array[DataType] = {
+    val outputSchema = ArrayBuffer[DataType]()
+    inputSchema.zipWithIndex.foreach {
+      // extract output type of explode from input ArrayData
+      case (dataType, index) if index == genOffset =>
+        require(dataType.isInstanceOf[ArrayType], "GpuExplode only supports ArrayData now")
+        if (includePos) {
+          outputSchema += IntegerType
+        }
+        outputSchema += dataType.asInstanceOf[ArrayType].elementType
+      // map types of other required columns
+      case (dataType, _) =>
+        outputSchema += dataType
+    }
+    outputSchema.toArray
+  }
+
+  override def fixedLenLazyExpressions: Seq[Expression] = child match {
+    // GpuLiteral of ArrayData will be converted to GpuCreateArray with GpuLiterals
+    case GpuCreateArray(expressions, _) => expressions
+    case GpuAlias(GpuCreateArray(expressions, _), _) => expressions
+    case _ => Nil
+  }
+
+  override def fixedLenLazyArrayGenerate(inputIterator: Iterator[ColumnarBatch],
+      boundLazyProjectList: Seq[Expression],
+      boundOthersProjectList: Seq[Expression],
+      outputSchema: Array[DataType],
+      outer: Boolean,
+      numOutputRows: GpuMetric,
+      numOutputBatches: GpuMetric,
+      totalTime: GpuMetric): Iterator[ColumnarBatch] = {
+
+    val numArrayColumns = boundLazyProjectList.length
+    val numOtherColumns = boundOthersProjectList.length
+    val numExplodeColumns = if (position) 2 else 1
+
+    new Iterator[ColumnarBatch] {
+
+      var currentBatch: ColumnarBatch = _
+      var indexIntoData = 0
+
+      private def closeCurrent(): Unit = if (currentBatch != null) {
+        currentBatch.close()
+        currentBatch = null
+      }
+
+      TaskContext.get().addTaskCompletionListener[Unit](_ => closeCurrent())
+
+      def fetchNextBatch(): Unit = {
+        indexIntoData = 0
+        closeCurrent()
+        if (inputIterator.hasNext) {
+          currentBatch = inputIterator.next()
+        }
+      }
+
+      override def hasNext: Boolean = {
+        if (currentBatch == null || indexIntoData >= numArrayColumns) {
+          fetchNextBatch()
+        }
+        currentBatch != null
+      }
+
+      override def next(): ColumnarBatch = {
+        if (currentBatch == null || indexIntoData >= numArrayColumns) {
+          fetchNextBatch()
+        }
+        withResource(new NvtxWithMetrics("GpuGenerateExec", NvtxColor.PURPLE, totalTime)) { _ =>
+          val result = new Array[ColumnVector](numExplodeColumns + numOtherColumns)
```

Comment:
nit: you can do a withResource on an Array too.

Reply:
Done.